### DISCUSSION TITLE
Has anyone encountered a "ReferenceError: document is not defined" when using Annotorious 3.x with OpenSeaDragon and Next.js?

### DISCUSSION THREAD

**Discussion by Songhao-LI** [annotorious/annotorious#455] [Q&A]
I'm using the following versions in my project:

Next.js: ^14.2.13
React: 18.2.0
@annotorious/react: ^3.0.11
@annotorious/openseadragon: ^3.0.11
When I try to render the page, I encounter the following error during server-side rendering (SSR):

⨯ ReferenceError: document is not defined
    at C:\Users\lsoho\Git\MyProject\node_modules\openseadragon\build\openseadragon\openseadragon.js:970:29
    at C:\Users\lsoho\Git\MyProject\node_modules\openseadragon\build\openseadragon\openseadragon.js:973:6
    at Object.<anonymous> (C:\Users\lsoho\Git\MyProject\node_modules\openseadragon\build\openseadragon\openseadragon.js:1071:2)
    at Module._compile (node:internal/modules/cjs/loader:1369:14)
    at Module._extensions..js (node:internal/modules/cjs/loader:1427:10)
    at Module.load (node:internal/modules/cjs/loader:1206:32)
    at Module._load (node:internal/modules/cjs/loader:1022:12)
    at cjsLoader (node:internal/modules/esm/translators:366:17)
    at ModuleWrap.<anonymous> (node:internal/modules/esm/translators:315:7)
    at ModuleJob.run (node:internal/modules/esm/module_job:222:25)
    at async ModuleLoader.import (node:internal/modules/esm/loader:323:24)

And here's my Code:
![image](https://github.com/user-attachments/assets/54ff91ab-8bbc-463d-926b-6b8dec7075c4)

When I init a new Next.js project without any other packages, I still meet the issue

![image](https://github.com/user-attachments/assets/90436c1d-b1bd-4e88-bbf4-657540241861)
![image](https://github.com/user-attachments/assets/1f43a163-1f03-4955-aa39-14d2f6c7bd60)

**Comment by rsimon** [comment]
Yes, I have seen the problem.

First: make absolutely sure that any OpenSeadragon- or Annotorious-related code runs only in the browser, never in SSR!

Even then: unfortunately, OpenSeadragon does seem to access `document` on the global level, even before initializing the viewer. Which means that this error can happen (on an SSR-based framework) as soon as you `import` OpenSeadragon. 

To repeat: this is __not caused by Annotorious__, but OpenSeadragon (see the error message). It happens as soon as OpenSeadragon gets imported as a dependency. (Note that even if you don't import OpenSeadragon yourself, it gets imported as a transitive dependency as soon as you import `@annotorious/openseadragon` or `@annotorious/react`.)

Below is how I worked around this limitation in an Astro project, using an async import. However I believe that Next has better utilities to "mark" libraries that only work in the browser and skip importing them in SSR accordingly.

https://github.com/annotorious/annotorious-website/blob/main/src/components/OpenSeadragonExample.tsx#L11-L18

```
const {
    Annotorious,
    OpenSeadragonAnnotator,
    OpenSeadragonViewer,
    ShapeType,
    useAnnotator,
    useSelection
  } = await import('@annotorious/react');
```

**Reply by Songhao-LI** [reply]
Really helpful!!!

**Reply by Songhao-LI** [reply]
Would you mind if sharing the code how you implement this part so as to help me better understand how to use annotorious 3.x. @rsimon 
![image](https://github.com/user-attachments/assets/6e620b26-ebc3-45c4-b0de-6e5a802ff476)

I am having an issue where I can't start drawing, even when holding the Shift key
![image](https://github.com/user-attachments/assets/f5ca27a7-15f9-4c88-8b7d-637a56f453a2)
![image](https://github.com/user-attachments/assets/eb7c63a2-6f36-4b86-a5a8-d783b2118fb6)

**Reply by rsimon** [reply]
You're missing the CSS stylesheet. See step 2 of the step-by-step guide here: https://annotorious.dev/react/openseadragon-iiif/

Note that you don't need to press the Shift key with Annotorious v3.

The code for the docs Website is here: https://github.com/annotorious/annotorious-website/blob/main/src/components/OpenSeadragonExample.tsx

**Reply by Songhao-LI** [reply]
Sorry for asking too much!! Have you ever encountered the error when packaging the project? @rsimon 

Everything works fine when I run the project locally, and I've already used dynamic imports to resolve a previous issue.
However, when packaging the project, I encountered this error:
![image](https://github.com/user-attachments/assets/4c3c9f0b-49d2-4d40-90f2-3a04190e8f96)

And here's how I import the packages:
![image](https://github.com/user-attachments/assets/e674f1a8-1a3b-4db5-b5af-1154a79a4e7d)

**Reply by Songhao-LI** [reply]
solved!!
![image](https://github.com/user-attachments/assets/8f09a2da-1025-48eb-a07b-063e4ffeec39)

**Reply by Songhao-LI** [reply]
But it seems like we still need to modify the library by adding this, it there any better solutions? @rsimon 

![image](https://github.com/user-attachments/assets/c4494b10-a4f4-4a23-bc50-c09f2913fe41)

**Comment by Clasen00** [comment]
Solved this isue like that:
```
const PreparationCanvas = dynamic(
  () =>
    import('../PreparationCanvas/PreparationCanvas').then(
      (mod) => mod.PreparationCanvas
    ),
  { ssr: false }
);
```
where PreparationCanvas is my Annotorious component. No  problem at all
